### PR TITLE
Fix Python and R repository smoke tests

### LINKED ISSUE #2107: Smoke tests failing on Tier 3 equivalent SRE
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **deploying** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** macOs/Linux
- **Data Safe Haven version:** release-v5.0.0rc3

## :package: Packages

<details>
<summary>List of packages</summary>

```none
Paste list of packages here
```

</details>

## :no_entry_sign: Describe the problem

Smoke tests fail on Tier 3 SREs because necessary packages are not on the current allowlists. The [allowlists](https://github.com/alan-turing-institute/data-safe-haven/tree/7b4e6f464e0fec4f7a1461a475d9032b0c9101cc/data_safe_haven/resources/software_repositories/allowlists) need to be brought up-to-date.

## :deciduous_tree: Log messages

<details>
<summary>Relevant log messages</summary>

```none
root@shm-turtles-sre-leonardo-vm-workspace-01:/usr/local/smoke_tests# bats run_all_tests.bats 
 ✓ Mounted drives (/data) 
 ✓ Mounted drives (/home) 
 ✓ Mounted drives (/output) 
 ✓ Mounted drives (/shared) 
 ✗ Python package repository
   (from function `initialise_python_environment' in file run_all_tests.bats, line 11,
    in test file run_all_tests.bats, line 66)
     `initialise_python_environment' failed
     ERROR: HTTP error 403 while getting http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl#sha256=2cd581cf58ab7fcfca4ce8efa6dcacd0de5bf8d0a3eb9ec927e07405f4d9e2a2 (from http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/simple/pip/) (requires-python:>=3.8)
   ERROR: Could not install requirement pip from http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl#sha256=2cd581cf58ab7fcfca4ce8efa6dcacd0de5bf8d0a3eb9ec927e07405f4d9e2a2 because of HTTP error 403 Client Error: Forbidden for url: http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl for URL http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl#sha256=2cd581cf58ab7fcfca4ce8efa6dcacd0de5bf8d0a3eb9ec927e07405f4d9e2a2 (from http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/simple/pip/) (requires-python:>=3.8)
 ✗ R package repository
   (in test file run_all_tests.bats, line 73)
     `[ "$status" -eq 0 ]' failed with status 2
 ✗ Python functionality
   (from function `initialise_python_environment' in file run_all_tests.bats, line 11,
    in test file run_all_tests.bats, line 80)
     `initialise_python_environment' failed
     ERROR: HTTP error 403 while getting http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl#sha256=2cd581cf58ab7fcfca4ce8efa6dcacd0de5bf8d0a3eb9ec927e07405f4d9e2a2 (from http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/simple/pip/) (requires-python:>=3.8)
   ERROR: Could not install requirement pip from http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl#sha256=2cd581cf58ab7fcfca4ce8efa6dcacd0de5bf8d0a3eb9ec927e07405f4d9e2a2 because of HTTP error 403 Client Error: Forbidden for url: http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl for URL http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl#sha256=2cd581cf58ab7fcfca4ce8efa6dcacd0de5bf8d0a3eb9ec927e07405f4d9e2a2 (from http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/simple/pip/) (requires-python:>=3.8)
 ✗ R functionality
   (from function `install_r_package_version' in file run_all_tests.bats, line 31,
    in test file run_all_tests.bats, line 87)
     `install_r_package_version "MASS" "7.3-52"' failed
   trying URL 'http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/cran-proxy/src/contrib/remotes_2.5.0.tar.gz'
   Error in download.file(url, destfile, method, mode = "wb", ...) : 
     cannot open URL 'http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/cran-proxy/src/contrib/remotes_2.5.0.tar.gz'
   In addition: Warning message:
   In download.file(url, destfile, method, mode = "wb", ...) :
     cannot open URL 'http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/cran-proxy/src/contrib/remotes_2.5.0.tar.gz': HTTP status was '403 Forbidden'
   Warning in download.packages(pkgs, destdir = tmpd, available = available,  :
     download of package ‘remotes’ failed
   Error in library("remotes", lib = "/root/.local/bats-r-environment") : 
     there is no package called ‘remotes’
   Execution halted
 ✗ MS SQL database (Python)
   (from function `initialise_python_environment' in file run_all_tests.bats, line 11,
    in test file run_all_tests.bats, line 98)
     `initialise_python_environment' failed
     ERROR: HTTP error 403 while getting http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl#sha256=2cd581cf58ab7fcfca4ce8efa6dcacd0de5bf8d0a3eb9ec927e07405f4d9e2a2 (from http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/simple/pip/) (requires-python:>=3.8)
   ERROR: Could not install requirement pip from http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl#sha256=2cd581cf58ab7fcfca4ce8efa6dcacd0de5bf8d0a3eb9ec927e07405f4d9e2a2 because of HTTP error 403 Client Error: Forbidden for url: http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl for URL http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl#sha256=2cd581cf58ab7fcfca4ce8efa6dcacd0de5bf8d0a3eb9ec927e07405f4d9e2a2 (from http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/simple/pip/) (requires-python:>=3.8)
 ✓ MS SQL database (R) 
 ✗ Postgres database (Python)
   (from function `initialise_python_environment' in file run_all_tests.bats, line 11,
    in test file run_all_tests.bats, line 115)
     `initialise_python_environment' failed
     ERROR: HTTP error 403 while getting http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl#sha256=2cd581cf58ab7fcfca4ce8efa6dcacd0de5bf8d0a3eb9ec927e07405f4d9e2a2 (from http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/simple/pip/) (requires-python:>=3.8)
   ERROR: Could not install requirement pip from http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl#sha256=2cd581cf58ab7fcfca4ce8efa6dcacd0de5bf8d0a3eb9ec927e07405f4d9e2a2 because of HTTP error 403 Client Error: Forbidden for url: http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl for URL http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/packages/pip/24.2/pip-24.2-py3-none-any.whl#sha256=2cd581cf58ab7fcfca4ce8efa6dcacd0de5bf8d0a3eb9ec927e07405f4d9e2a2 (from http://nexus.leonardo.blue.develop.turingsafehaven.ac.uk/repository/pypi-proxy/simple/pip/) (requires-python:>=3.8)
 ✓ Postgres database (R) 

12 tests, 6 failures
```

</details>

## :recycle: To reproduce

Run smoke tests on a Tier 3 SRE

### PR DESCRIPTION
**Pull request by craddm**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

This is best done in combination with an update to the `nexus-allowlist` container to allow installation of older versions of R packages. Current smoke tests will still fail as they require installation of older R package versions, which is not currently allowed by the Nexus server.

### :arrow_heading_up: Summary

Minimally updates the PyPi allowlist, ensuring that the smoke tests can run with all required dependencies now on the list.

Updates the R smoke tests to now refer to packages that are on allowlist.

Fixes the following tests:
- Python package repository
- R package repository
- Python functionality
- MS SQL database (Python & R)
- Postgres database (Python & R)

Does not fix the R functionality test.

### :closed_umbrella: Related issues

Closes #2107 

### :microscope: Tests

Tested on a Tier 3 equivalent SRE deployment - `noteasybeing.green.develop.turingsafehaven.ac.uk`

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/release_checklist.md [hunk 1]
```diff
@@ -25,12 +25,10 @@ Refer to the [Deployment](https://data-safe-haven.readthedocs.io/en/latest/deplo
 ### For minor releases and above
 
 - [ ] Deploy an SHM from this branch and save a transcript of the deployment logs
-- [ ] Build an SRE compute image (SRD) and save transcripts of the logs
 - Using the new image, deploy a tier 2 and a tier 3 SRE
     - [ ] Save the transcript of your tier 2 SRE deployment
     - [ ] Save the transcript of your tier 3 SRE deployment
 - [ ] Complete the [Security evaluation checklist](https://data-safe-haven.readthedocs.io/en/latest/deployment/security_checklist.html) from the deployment documentation
-- [ ] Update [SECURITY.md](../../SECURITY.md) and [VERSIONING.md](../../VERSIONING.md)
```
Comment (jemrobinson) [inline .github/ISSUE_TEMPLATE/release_checklist.md]: I think we might still want this line.

File: VERSIONING.md [hunk 2]
```diff
@@ -91,8 +91,9 @@ The following versions have been evaluated by third party security consultants p
...
 | [v4.0.0](https://github.com/alan-turing-institute/data-safe-haven/releases/tag/v4.0.0)          | 2 September 2022  | Penetration test evaluating ability to infiltrate/exfiltrate data from the system.                                                                                                                 | No major security issues identified. |
-| [v5.0.0-rc1](https://github.com/alan-turing-institute/data-safe-haven/releases/tag/v5.0.0-rc.1) | 18 September 2023 | Penetration test evaluating ability to infiltrate/exfiltrate data from the system. Testing next codebase, using Python and Pulumi.                                                                        | No major security issues identified. |
+| [v5.0.0-rc1](https://github.com/alan-turing-institute/data-safe-haven/releases/tag/v5.0.0-rc.1) | 18 September 2023 | Penetration test evaluating ability to infiltrate/exfiltrate data from the system. Testing next codebase, using Python and Pulumi.                                                                 | No major security issues identified. |
```
Comment (jemrobinson) [inline VERSIONING.md]: Can we add 5.0.0-rc2 here too? Or can do in another PR if preferred.

File: .github/ISSUE_TEMPLATE/release_checklist.md (hunk 1 above)
Comment (craddm) [reply .github/ISSUE_TEMPLATE/release_checklist.md]: I'm not sure how that change crept in there

File: VERSIONING.md (hunk 2 above)
Comment (craddm) [reply VERSIONING.md]: Right, this and the other one are actually changes Jim made in the `release-v5.0.0` branch, which this was based on, and I assume they're showing up because i put `develop` as the target branch instead of `release-v5.0.0`

**Comment by jemrobinson** [comment]
@craddm : do you want to merge this as a fix for Python/PyPI only or wait until you have a fix for R/CRAN too?

**Comment by craddm** [comment]
> @craddm : do you want to merge this as a fix for Python/PyPI only or wait until you have a fix for R/CRAN too?

Could do; smoke tests will still fail for R until `nexus-allowlists` is updated, and then we'll need a separate PR to update which version of that we're using. Or I can probably change the R smoke tests further so they pass now.

**Comment by jemrobinson** [comment]
I think it's better to keep PRs small if possible. Is this one working and ready for review?

**Comment by craddm** [comment]
I hadn't finished it, no, as I switched to getting the CRAN part working via `nexus-allowlist`, and I still need to work out why some of the smoke test are still failing.

The database smoke tests are all failing at the moment (python and R)

I suspect this is a different problem to so I'll transfer this to new issue/PR

```
 ✗ Postgres database (Python)
   (in test file run_all_tests.bats, line 118)
     `[ "$status" -eq 0 ]' failed
```

**Comment by jemrobinson** [comment]
I've changed the title of the PR to what I think it's doing. Can you paste a before/after of the Python tests with this change?

**Comment by craddm** [comment]
> I've changed the title of the PR to what I think it's doing. Can you paste a before/after of the Python tests with this change?

This wasn't correct, as it also fixes an R smoke test to use packages that are actually on the allowlist

**Review by jemrobinson — APPROVED** [review]
LGTM. Can you detail which smoke tests are fixed by this somewhere?

**Comment by craddm** [comment]
Note, this doesn't fix the R functionality test, as that requires installation of an old version of `MASS` that is compatible with our R version. That'd be fixed by #2116, or by changing to a different package.

**Comment by JimMadge** [comment]
@craddm Do you want to check if #2116 would fix this?

I think the conversation about whether than is fine to put in the release is still open.
If it makes the difference here I think we should try to include it.

**Comment by jemrobinson** [comment]
I don't think we can release with broken smoke tests. I doubt (although I can't be completely certain) that updating the `nexus-allowlists` version will trigger any vulnerability that wasn't already present. If we're worried about that, let's update one of the under-test SREs and ask the penetration testers to retest the software packages server.

**Comment by craddm** [comment]
I also think the ramifications of this change in terms of allowing more possible downloads from CRAN are not very concerning. It's still allowing R packages that have been through the same process as all the latest R packages, and only from CRAN, so as James says I don't think it triggers any vulnerability that isn't already there.

**Comment by JimMadge** [comment]
> I also think the ramifications of this change in terms of allowing more possible downloads from CRAN are not very concerning. It's still allowing R packages that have been through the same process as all the latest R packages, and only from CRAN, so as James says I don't think it triggers any vulnerability that isn't already there.

:+1: Moreover, I think these changes are making Nexus behave as we expected. We already decided in T2 we are comfortable with all CRAN packages, and in T3 we are not interesting in filtering versions.

**Comment by jemrobinson** [comment]
OK, so sounds like we should merge this PR and also test out #2116. Is that right?

**Comment by craddm** [comment]
I'm testing out #2116 shortly, in combination with this
